### DISCUSSION TITLE
DDEV Test Environment for OpenMage in Windows 10/11

### DISCUSSION THREAD

**Discussion by addison74** [OpenMage/magento-lts#3839] [Show & Tell]
**_Last update February 27, 2025_**

### VMware Workstation - Free for Personal Use

In this tutorial I used VMware Workstation as a desktop hypervisor. Starting with spring 2024, the new owner Broadcom has decided to provide the application with two license models, Free Personal Use or Paid Commercial use. So, you can use this application for free as long as it is for personal use. I encourage everyone to use VMware because compared to its competitors (VirtualBox, Microsoft Hyper-V) it has better performance especially on the graphics side. A Windows 10 or 11 operating system runs in a virtual machine pretty decent to a real one. 

For more details read this article
https://blogs.vmware.com/workstation/2024/05/vmware-workstation-pro-now-available-free-for-personal-use.html

### VMware Workstation Settings
If you run Windows OS in a virtual machine, it is mandatory to activate the virtualization option. For example, in VMware with the virtual machine off, access `Edit virtual machine settings`. In the `Hardware` tab select `Processors` and in the `Virtualization engine` section check the `Virtualize Intel VT-x/EPT or AMD-V/RVI` option.

You need at least 8 GB of memory to run (Docker + PHPStorm) decent. I recommend a machine with 16 GB.

![windows_1909](https://github.com/OpenMage/magento-lts/assets/8360474/33c7605e-b9ff-46a5-b960-0e2aabb4dc77)

### (Windows) Installing the Windows Terminal Application
1. Open the `Microsoft Store` application and search for `Windows Terminal`
2. Choose the first result then press the `Get` button

![windows_terminal](https://github.com/OpenMage/magento-lts/assets/8360474/fbcb31fa-3582-4372-8acd-48f4956d6d30)

If you use PHPStorm, at the bottom there is a tab called `Terminal`. Here you can run commands in the Linux distribution too.

![phpstorm_terminal](https://github.com/OpenMage/magento-lts/assets/8360474/cca5b48d-dd5b-4d80-8664-a4f99ee9f960)

### (Windows) Installing WSL2 (Windows Subsystem for Linux 2)
**The Installation Tutorial**
https://pureinfotech.com/install-windows-subsystem-linux-2-windows-10/

**Useful Commands**
https://learn.microsoft.com/en-us/windows/wsl/
```
wsl --install
wsl --version
wsl --update
wsl --list --online
wsl --install -d DISTRO-NAME
wsl --set-version <distro name> 2
```

**Advanced Settings**
https://learn.microsoft.com/en-us/windows/wsl/wsl-config#configure-global-options-with-wslconfig

You can configure limits on the memory, CPU and swap size allocated to WSL 2 in a `.wslconfig` file⁠. For example, create a file named `.wslconfig` in the `C:\Users\<User Name>` directory with the following content

```
[wsl2]
memory=4GB   # Limits VM memory in WSL 2 up to 4 GB
processors=6 
```

### (Windows) Installing a Linux Distribution
**Variant 1**
1.  Open the `Command Prompt` application as Administrator
2. Get the distributions list `wsl --list --online`
3. Install a distribution `wsl --install -d Ubuntu-20.04`
4. Reboot the system

**Variant 2**
1.  Open the `Microsoft Store` application and search for Ubuntu
2. Install an LTS version, for example Ubuntu 20.04.6 LTS
3. Reboot the system

Open the `Terminal` application then choose a new tab with Ubuntu. Follow the steps to complete the installation, setting the `username` and `password` (e.g. ubuntu / 1234).

![terminal](https://github.com/OpenMage/magento-lts/assets/8360474/b88fe7ae-4d86-4cba-a239-01c06a685bec)

### (Windows) Installing Docker
https://docs.docker.com/desktop/install/windows-install/

1. Download the installation file `Docker Desktop Installer.exe` and run it as Administrator
2. During installation check the option `Use WSL 2 instead of Hyper-V (recommended)`
3. Reboot the system

### (Windows) Installing mkcert for Secured Connections
https://github.com/FiloSottile/mkcert

1. Download the latest Windows release
2. Open the `Terminal` application as Administrator
3. Go to the directory where you downloaded the executable file named `mkcert-vX.X.X-windows-amd64.exe`
4. Run `mkcert-vX.X.X-windows-amd64.exe --install`
5. In the popup window, where you are asked if you want to install the certificate, press `Yes`

### (Windows) Linux distribution drive mapping

1. Open the `Windows Explorer` application
2. On the left side open the `Linux` path and select `Ubuntu-20.04`
3. Right-click and select `Map network drive...` from the menu. Choose a letter, for example Z:

From now on it appears in the `This PC` section under `Network locations`. If you want to disconnect it, right-click and select `Disconnect` from the menu.

![windows_explorer](https://github.com/OpenMage/magento-lts/assets/8360474/97dacd28-a316-4312-8f5a-1eb67b796d07)

### (Windows) PHPStorm

Create a new project in PHPStorm that has the location where you cloned the OpenMage repository. If you log in to your GitHub account, you will see in the `Pull Request` tab on the left the open PRs from OpenMage. Open one and checkout to start testing. When you are done, at the bottom of the PHPStorm window you will see the `Git` tab. Click on it, then right click on `Local > main` and select `Checkout`. You can update the repositories, locals and remotes, from time to time and more.

![new_project](https://github.com/OpenMage/magento-lts/assets/8360474/b408b23c-0128-4887-abc8-30e7133e4fb3)

### (Linux) Installing DDEV
https://ddev.readthedocs.io/en/latest/users/install/ddev-installation/

1. Open the `Terminal` application and choose a new tab with Ubuntu
2. Bring all the packages up to date Ubuntu `sudo apt update && sudo apt upgrade -y`
3. Install DDEV according to the instructions

### (Linux) Copying mkcert Certificates from Windows to Linux

**Variant 1**
1. Open the `Terminal` application and choose a new tab with Ubuntu
2. Create the path  `mkdir -p /home/<user_name/.local/share/mkcert`
3. Run `cp /mnt/c/Users/<User Name>/AppData/Local/mkcert/* /home/<user_name>/.local//share/mkcert/`

**Variant 2**
1. Run the `Windows Explorer` application as administrator
2. Copy the files `rootCA.pem` and `rootCA-key.pem` from `C:\Users\<User Name>\AppData\Local\mkcert`
3. Create the path in Linux > Ubuntu-20.04 `/home/<user_name/.local/share/mkcert`
4. Paste them in the `mkcert` directory

### (Linux) Installing OpenMage
First make sure that the `Docker Desktop` application is running in Windows.

1. Open the `Terminal` application and choose a new tab with Ubuntu
2. Create the path `mkdir -p /home/<user_name>/openmage`
3. Clone the OpenMage repository `git clone https://github.com/OpenMage/magento-lts.git /home/<user_name>/openmage`
4. Go to the directory `/home/<user_name>/openmage`
5. Run the following DDEV commands. The first command configures the project, the second installs OpenMage dependencies, the third installs Magento Sample Data.

```
ddev config
ddev composer install
ddev openmage-install -s -k
```

6. Edit the `.ddev/config.yaml` file to change the webserver and PHP version as you want. I am using Apache and PHP 8.3

```
php_version: "8.3"
webserver_type: apache-fpm
```

7. Run the following DDEV commands to load the project in the browser window with a secured connection

```
ddev start
ddev launch
```

### (WSL) The xdg-open error when running the ddev launch command
If you encounter errors when running the ```ddev launch``` command, related to xdg-open and a long list of missing Linux browsers it is normal. DDEV wants to launch an installed browser in WSL and since there is none, we have to configure to use the default one in Windows. All you have to do is editing the **~/.bashrc** file and adding the following line to the end

```export BROWSER="powershell.exe /C start"```

For the change to take effect, run ```. .bashrc.```. 

Another variant would be to install in WSL the **wslu package**.

### Conclusion
As you can see, the more complicated part is the initial configuration of WSL, Docker, DDEV. Once done, it doesn't take more than 3-5 minutes to get an instance of OpenMage ready for testing. Forget about XAMPP, WAMP in Windows. The free nature of VMware Workstation allows you to move a virtual machine into many desktops and have a ready to go testing environment.

You can use a test environment in Windows without WSL2, but I do not recommend this configuration because DDEV must be installed in Windows and not inside the Linux distribution. It depends on Mutagen and it is very very slow. Testing in a Linux distribution has many advantages and it is close to the final stage, when you have to move the project into production.

For more information about using DDEV please visit https://docs.openmage.org/developers/tools/ddev/. It is a fantastic tool!

### Proxy Connection
When you connect to the Internet through a proxy connection, with or without authentication, because the network administrator has banned the access to certain websites (YouTube, Facebook, TikTok), you have to make some changes to be able to run properly the DDEV-based test environment. Please follow the tutorial I created here https://github.com/OpenMage/magento-lts/discussions/4656.

**Comment by addison74** [comment]
Those who use PHPStorm as IDE can integrate DDEV in the editor.There is already a **plugin** that offers more features. For more details visit

https://plugins.jetbrains.com/plugin/18813-ddev-integration

**Comment by kiatng** [comment]
Question on _(Windows) Installing mkcert for Secured Connections_
So, I downloaded mkcert following the `scoop` instructions on https://github.com/FiloSottile/mkcert :

```bash
scoop bucket add extras
scoop install mkcert
```

I cannot connect my steps with this step:

> 2. Go to the directory where you downloaded the executable file named mkcert-vX.X.X-windows-amd64.exe

Where do I look for the file `mkcert-vX.X.X-windows-amd64.exe`? What am I missing?

[edit] Got it, the windows executable to install mkcert can be downloaded here: https://github.com/FiloSottile/mkcert/releases

**Reply by addison74** [reply]
You don't have to use scoop to install mkcert. Just download a release then run the executable in a terminal window, with administrator rights, as mentioned in the tutorial.

**Comment by kiatng** [comment]
I encountered the following error:
```bash
kiat@DESKTOP-7GLIBDK:~/openmage$ ddev composer install
Building project images...
Project images built in 1s.
 Container ddev-openmage-db  Recreate
 Container ddev-openmage-web  Recreate
 Container ddev-openmage-db  Recreated
 Container ddev-openmage-web  Recreated
 Container ddev-openmage-db  Started
 Container ddev-openmage-web  Started
Failed to create certificates for project, check mkcert operation: ERROR: failed to read the CA certificate: unexpected content
; err=exit status 1
```

What do I do?

**Reply by addison74** [reply]
In **WSL2 + Ubuntu/Debian**, the following important steps must be followed, as I mentioned in the tutorial

1. Installing DDEV
2. Copying mkcert Certificates from Windows to Linux

According to the documentation, https://ddev.readthedocs.io/en/latest/users/install/ddev-installation/, installing DDEV in Linux requires creating certificates with mkcert. You have all the steps in the documentation, in the tab named Linux. Then you copy the certificates created by mkcert in Windows over those in WSL2.

Please do the following

1. Check if mkcert was installed in WSL2 and if you ran the ```mkcert --install``` command. Two PEM files will be generated here ```~/.local/share/mkcert/```.

2. If you copied & pasted the rootCA*.pem files from Windows to WSL2, check the permissions in the ```~/.local/share/mkcert/``` directory to make sure they don't have root permissions.

If the PEM files do not exist, the following message is displayed

```console
mkcert may not be properly installed, we suggest installing it for trusted https support, `brew install mkcert nss`, `choco install -y mkcert`, etc. and then `mkcert -install`
```

In your case, it must be checked if those files exist, if they are the ones generated in Windows and the permissions (for example ubuntu ubuntu). If it is not working check this posts

https://github.com/ddev/ddev/issues/2415
https://github.com/ddev/ddev/issues/5536
https://github.com/FiloSottile/mkcert/issues/557

It seems older mkcert versions had some issues. If you did no install it by download then ```dpkg -i .... ``` this could be the problem.

There is a chance that the certificates generated in Windows do not have the correct format, maybe because you used scoop. mkcert executables from Windows and Ubuntu must be the same version. Manual installation via download will guarantee that. 

Following the steps in the tutorial I posted would not have created any problems for you because. I re-checked them a few moments ago and everything is fine.

**Reply by kiatng** [reply]
Thanks for taking the time to re-check everything. I tried all kinds of commands, I do not know what I am doing, but finally the following steps worked:

**Step 1**: I deleted files `rootCA-key.pem` and  `rootCA.pem` from `C:\Users\user1\AppData\Local\mkcert` and `~/.local/share/mkcert`. 
```bash
kiat@DESKTOP-7GLIBDK:~/.local/share/mkcert$ rm root*
rm: remove write-protected regular file 'rootCA-key.pem'? y 
kiat@DESKTOP-7GLIBDK:~/.local/share/mkcert$ ll
total 8
drwxr-xr-x 2 kiat kiat 4096 Mar 11 09:18 ./
drwx------ 4 kiat kiat 4096 Mar 10 13:04 ../
```
For the one in windows, I needed to delete the 2 files in Windows Explorer. Deleting in Administrator:PowerShell encountered an error:
```powershell
PS C:\Users\user1\AppData\Local\mkcert> Remove-Item root*
Remove-Item : Cannot remove item C:\Users\user1\AppData\Local\mkcert\rootCA-key.pem: You do not have sufficient access rights to perform this operation.
At line:1 char:1
+ Remove-Item root*
+ ~~~~~~~~~~~~~~~~~
    + CategoryInfo          : PermissionDenied: (C:\Users\user1\...\rootCA-key.pem:FileInfo) [Remove-Item], IOException
    + FullyQualifiedErrorId : RemoveFileSystemItemUnAuthorizedAccess,Microsoft.PowerShell.Commands.RemoveItemCommand
```

**Step 2**: In Ubuntu terminal, I uninstall something:

```bash
kiat@DESKTOP-7GLIBDK:~/.local/share/mkcert$ mkcert -uninstall
Created a new local CA 💥
Sudo password:
The local CA is now uninstalled from the system trust store(s)! 👋
```

That was unexpected, because previously, the result was always
```bash
kiat@DESKTOP-7GLIBDK:~/.local/share/mkcert$ mkcert -uninstall
ERROR: failed to read the CA certificate: unexpected content
```

It also created the 2 files `rootCA-key.pem` and  `rootCA.pem` in `C:\Users\user1\AppData\Local\mkcert` but not in `~/.local/share/mkcert`.

**Step 3**:  I did this:
```bash
kiat@DESKTOP-7GLIBDK:~/openmage$ ddev composer install
Building project images...
Project images built in 1s.
 Container ddev-openmage-web  Recreate
 Container ddev-openmage-db  Recreate
 Container ddev-openmage-db  Recreated
 Container ddev-openmage-web  Recreated
 Container ddev-openmage-web  Started
 Container ddev-openmage-db  Started
Waiting for web/db containers to become ready: [web db]
Starting ddev-router if necessary...
 Container ddev-router  Created
 Container ddev-router  Started
Waiting for additional project containers to become ready...
All project containers are now ready.
```
I didn't expect it to work as I didn't copy the 2 files `rootCA-key.pem` and  `rootCA.pem` in `C:\Users\user1\AppData\Local\mkcert` to `~/.local/share/mkcert`. But `ddev composer install` didn't complain.

**Reply by kiatng** [reply]
Continue...

**Step 4**: In Ubuntu terminal, I issued command `ddev openmage-install -s -k`:

```bash```
kiat@DESKTOP-7GLIBDK:~/openmage$ ddev openmage-install -s -k
Enable maintenance mode...
OpenMage is already installed. Delete local.xml and drop the database? [y/N] y
Creating Sample Data directory...
Downloading Sample Data...
--2024-03-11 01:54:19--  https://github.com/Vinai/compressed-magento-sample-data/raw/master/compressed-magento-sample-data-1.9.2.4.tgz
Resolving github.com (github.com)... 20.205.243.166
Connecting to github.com (github.com)|20.205.243.166|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/Vinai/compressed-magento-sample-data/master/compressed-magento-sample-data-1.9.2.4.tgz [following]
--2024-03-11 01:54:19--  https://raw.githubusercontent.com/Vinai/compressed-magento-sample-data/master/compressed-magento-sample-data-1.9.2.4.tgz
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.111.133, 185.199.108.133, 185.199.109.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.111.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 68614817 (65M) [application/octet-stream]
Saving to: ‘sample_data.tgz’

sample_data.tgz                      100%[======================================================================>]  65.44M  41.1MB/s    in 1.6s

2024-03-11 01:54:26 (41.1 MB/s) - ‘sample_data.tgz’ saved [68614817/68614817]

Uncompressing Sample Data...
Copying Sample Data into the OpenMage directory...
Clearing var/cache...
Importing Sample Data into the database...
Admin User [admin]:
Admin Firstname [OpenMage]:
Admin Lastname [Administrator]:
Admin Email [admin@example.com]:
Admin Password [veryl0ngpassw0rd]:
SUCCESS: 76b6192b2b8806b5d574e94c63dd6a0c
Disable maintenance mode...
kiat@DESKTOP-7GLIBDK:~/openmage$
```

I used default values for everything.

**Step 5**: In VsCode, I edited the file `Z:\home\kiat\openmage\.ddev\config.yaml`, I changed only `php_version: "8.3"` from 7.4. 

**Step 6**: In VsCode, I open a Ubuntu terminal and issued `ddev start`

```bash
kiat@DESKTOP-7GLIBDK:~/openmage$ ddev start

 TIP OF THE DAY                                                                                                                                
 Lots of database browsers work with DDEV, including TablePlus, Sequel Pro, and DBeaver.                                                       
 https://ddev.readthedocs.io/en/stable/users/usage/database-management/#database-guis                                                          

Starting openmage... 
Building project images... 
.Project images built in 5s. 
 Container ddev-openmage-db  Recreate 
 Container ddev-openmage-web  Recreate
 Container ddev-openmage-db  Recreated 
 Container ddev-openmage-web  Recreated 
 Container ddev-openmage-db  Started 
 Container ddev-openmage-web  Started 
Waiting for web/db containers to become ready: [web db] 
Starting ddev-router if necessary... 
 Container ddev-router  Running 
Waiting for additional project containers to become ready... 
All project containers are now ready. 
/home/kiat/openmage/app/etc/local.xml already exists and is managed by the user.
Successfully started openmage
Project can be reached at https://openmage.ddev.site https://127.0.0.1:54981
```

**Step 7**: I encountered errors on `ddev launch`:

```bash
kiat@DESKTOP-7GLIBDK:~/openmage$ ddev launch
/usr/bin/xdg-open: 882: x-www-browser: not found
/usr/bin/xdg-open: 882: firefox: not found
/usr/bin/xdg-open: 882: iceweasel: not found
/usr/bin/xdg-open: 882: seamonkey: not found
/usr/bin/xdg-open: 882: mozilla: not found
/usr/bin/xdg-open: 882: epiphany: not found
/usr/bin/xdg-open: 882: konqueror: not found
/usr/bin/xdg-open: 882: chromium: not found
/usr/bin/xdg-open: 882: chromium-browser: not found
/usr/bin/xdg-open: 882: google-chrome: not found
/usr/bin/xdg-open: 882: www-browser: not found
/usr/bin/xdg-open: 882: links2: not found
/usr/bin/xdg-open: 882: elinks: not found
/usr/bin/xdg-open: 882: links: not found
/usr/bin/xdg-open: 882: lynx: not found
/usr/bin/xdg-open: 882: w3m: not found
xdg-open: no method available for opening 'https://openmage.ddev.site'
Failed to run launch ; error=exit status 3
```

But that doesn't matter because I can browse https://openmage.ddev.site/accessories.html in unsecure mode.
![image](https://github.com/OpenMage/magento-lts/assets/1106470/a71057e4-081e-4951-be6b-207b58d74e55)

So it seems I probably miss the step of copying the 2 files from window to Ubuntu. Not sure what I need to do from here to get secure browsing.

Because I know nothing, it has taken me many many hours over months (because I am busy with lots of other stuff) to reach here. I hope others have better experience.

**Reply by addison74** [reply]
1 - You must go through all the steps as listed in the tutorial. I would delete everything and start over. It is not complicated at all, after you repeat this process 2 times you won't need the instructions anymore.

2 - Regarding the **certificate**. You must use the **mkcert** program, both in Windows and Linux.

**(Windows)**
- run the ```mkcert -uninstall``` command. this uninstalls the certificate but doesn't delete the files
- delete the ```root*.pem``` files from the ```C:\Users\<Windows User>\AppData\Local\mkcert``` directory
- run the ```mkcert -install``` command. this installs the certificate and creates new files
- check if the certificate was installed in the application **Manage computer certificates** > Other People > Certificates

**(Linux)**
- run the ```mkcert -uninstall``` command. this uninstalls the certificate but doesn't delete the files
- delete the ```root*.pem``` files from ```/home/<Linux User>/.local/share/mkcert```
- run the ```mkcert -install``` command. this installs the certificate and creates new files

After doing these steps, if you run the commands in order ```ddev start```, ```ddev launch```, when the browser window opens automatically, you will see the message "Your connection is not private". This is happening because the browser uses the Linux certificate that is not installed in Windows. To solve this problem, you must copy the certificates from Windows over those from Linux

```
sudo cp /mnt/c/Users/<Windows User>/AppData/Local/mkcert/* ~/.local/share/mkcert/
ddev poweroff
ddev start
ddev launch
```
The security message is no longer displayed in the browser window. If it still appears, then run 
```
mkcert -install
ddev poweroff
ddev start
ddev launch
```

3 - The **/usr/bin/xdg-open** issue. This is not related to this tutorial, it's part of the WSL2 configuration. In the case of Ubuntu 20 I didn't have to do anything extra. You must check if you have installed the following packages **xdg-utils** and **wslu** in your distribution and if you don't have them, install them.

**Reply by kiatng** [reply]
I fixed the `ddev launch` issue by `$ sudo apt install wslu`, now it opens the Edge browser. However, I haven't gotten the _Not secure_ issue yet.

**Reply by kiatng** [reply]
I fixed the Not Secure browsing, ref [Discord](https://discordapp.com/channels/664580571770388500/1032287136818397194). It suggested to use the Firefox manual steps described  [here](https://ddev.readthedocs.io/en/latest/users/install/ddev-installation/#traditional-windows).  In my case, the first 2 steps

> Run `mkcert -install` (you can use the shortcut from the Start Menu for that)
> Run `mkcert -CAROOT` to see the local folder used for the newly-created root certificate authority

are aIl I need. After, I power-down ddev, then start and launch it. Apparently, somehow, I didn't do `mkcert -install` in Windows, or that I need to install it again.

**Reply by addison74** [reply]
If you like this development/testing environment, I recommend you to repeat the configuration from scratch several times to familiarize yourself with each step. On a new machine, it doesn't take me more than 15 minutes to configure everything I need and the OpenMage part it takes only a few commands. The tutorial can be fully applied in a Linux distribution, certain steps will no longer be necessary.

Now that you made it works, it's time to familiarize yourself with the DDEV commands. You can install phpmyadmin, cron and many others. Mailpit catches everything that is sent by email. DDEV doesn't only help me with OpenMage, I have several WordPress projects that I am working on. I did not test it with Magento 2.4 yet.

**Reply by kiatng** [reply]
15 minutes for you is 15 hours for me. Anyway, thanks for your help, without your persistency, I'd given up already.

**Reply by addison74** [reply]
After a while you will find that you did not waste your time. It would have been a loss to give up. Everyone who wants to give OpenMage a chance and wants to invest minimal effort in testing has this tutorial at their disposal.

**Comment by kiatng** [comment]
I tried again today. There is a message in VSCode to update DDEV. I did it with `sudo apt-get update && sudo apt-get upgrade -y` in WSL2 Ubuntu terminal. After verifying the ver with `ddev --version`, I issue the following in VSCode terminal:

```bash
$ ddev --version
ddev version v1.23.4
$ ddev start
$ ddev launch
```
I get 403 Forbidden in the browser:
![image](https://github.com/user-attachments/assets/9b2f3f82-c25a-431f-a13f-515a339596b6)

.ddev/config.yaml
```yml
name: openmage
type: magento
docroot: web
php_version: "8.3"
webserver_type: apache-fpm
xdebug_enabled: false
additional_hostnames: []
additional_fqdns: []
database:
    type: mariadb
    version: "10.4"
use_dns_when_possible: true
composer_version: "2"
web_environment: []
corepack_enable: false
default_container_timeout: "240"
```
Any idea what I did wrong?

**Reply by addison74** [reply]
I reviewed this tutorial and installed a test environment on a VM. It works as expected. If you followed all the steps you should not encounter any problem.

**I would check the following**

1. Run the ```ddev start``` command in the OpenMage root directory. This to ensure that you are in the right directory configured by DDEV. From your message I understand that the website loads in the browser, so you have a valid configuration, but it doesn't hurt to check.

2. For sure your problem comes from this line in the DDEV configuration file ```.ddev/config.yaml```
```php
docroot: web
```
Change it to
```
docroot: ""
```
then run the ```ddev restart``` command. If you don't have this directory called web, you basically ask the web server to load a root directory that doesn't exist and this is the reason you get the 403 error.

**Reply by sreichel** [reply]
If "web" is set by default I'll open an issue at DDEV. It should be empty.

**Reply by addison74** [reply]
I deleted the DDEV configuration file then I used the ```ddev config``` command again. The web value is an empty string, as I pointed above. 

@sreichel - it is not an issue in DDEV.

**Reply by kiatng** [reply]
Thanks @ADDISON74 `docroot: ""` this works. I also deleted the web folder in the root. It had local.xml in the app dir.

**Reply by sreichel** [reply]
Welcome to DDEV  :)

**Comment by kiatng** [comment]
For PHPMyAdmin, it can be invoked simply with `ddev phpmyadmin`, which will install it the first time the command is run, and launch it in the browser.

**Reply by sreichel** [reply]
Useful command ... `ddev describe` to see all available tools/services.

**Reply by addison74** [reply]
I guess starting with v1.23.4 there is no need to install phpMyAdmin first using ```ddev get ddev/ddev-phpmyadmin```.

If we missed something we can update it here https://github.com/OpenMage/magento-lts/blob/main/docs/DDEV.md.

**Reply by sreichel** [reply]
... or here https://github.com/OpenMage/magento-lts/pull/4156

**Reply by sreichel** [reply]
https://github.com/OpenMage/magento-lts/pull/4349/commits/6607b0d5afa08712bb52532e836ccce7ea8c8805

**Comment by sreichel** [comment]
@ADDISON74 can you add it to https://github.com/OpenMage/magento-lts/pull/4156 ?

**Reply by addison74** [reply]
Please add it as you want then I can evaluate it.

What's funny is that I made similar tutorials for OpenCart and Prestashop that were quite appreciated. DDEV really solves many problems, I discovered it thanks to you.

**Comment by addison74** [comment]
Small updates to this article

- VMware Workstation is free to use, no more payment
- Solving the `ddev launch` command error, the browser is not opening

**Comment by addison74** [comment]
I will create a separate article for the situation when the computer on which the DDEV test environment is installed is behind a proxy server with authentication. I encountered this situation recently where a corporation allows employees to access the Internet through a proxy server. Now it seems to me easy to implement, but initially it created some issues because the settings have to be made in both Windows and WSL for it to work.

**Reply by addison74** [reply]
The tutorial can be found here https://github.com/OpenMage/magento-lts/discussions/4656.
